### PR TITLE
Jk/cumulus 3518

### PR DESCRIPTION
**Pull request by Jkovarik**
**Summary:** Summary of changes

Addresses [CUMULUS-3518: Develop amazing new feature](https://bugs.earthdata.nasa.gov/browse/CUMULUS-3518)

## Changes

- **CUMULUS-3518**
  - Update existing usage of `@cumulus/aws-client` lambda service to use AWS SDK v3 `send` syntax

## PR Checklist

- [x] Update CHANGELOG
- [x] Unit tests
- [ ] Ad-hoc testing - Deploy changes and test manually
- [ ] Integration tests

### CODE REVIEW

**Comment by Jkovarik** [comment]
PR is still WIP as I'm tracking down some instability in the CI units.     Main code effort should be reviewable however.

File: example/cumulus-tf/variables.tf [hunk 1]
```diff
@@ -424,7 +424,7 @@ variable "lambda_memory_sizes" {
   type = map(number)
   default = {
     cleanExecutions = 384           # archive
-    DistributionApiEndpoints = 384  # cumulus_distribution
+    DistributionApiEndpoints = 512  # cumulus_distribution
```
Comment (npauzenga) [inline example/cumulus-tf/variables.tf:427]: Might be worth breaking out into a separate PR depending on timing. Fine to leave it here too.

File: example/package.json [hunk 2]
```diff
@@ -72,6 +72,7 @@
   },
   "devDependencies": {
     "@aws-sdk/client-ecs": "^3.447.0",
+    "@aws-sdk/client-lambda": "^3.529.1",
```
Comment (npauzenga) [inline example/package.json:75]: I think these should be regular `dependencies` instead of `devDependencies`. In my PR I moved client-ecs up. It's a little weird because these are exclusively(?) used in integration tests in the `example` package. Thoughts?

File: packages/api-client/tests/test-CumulusApiClient.js [hunk 3]
```diff
@@ -26,8 +26,8 @@ test.serial('invokeApi invokes the lambda with the expected Payload and Function
   };
 
   fakeServices.lambda = () => ({
-    invoke: (payloadObject) => {
-      const passedPayload = payloadObject;
+    send: (payloadObject) => {
```
Comment (npauzenga) [inline packages/api-client/tests/test-CumulusApiClient.js:29]: Hmm different approach to checking that the lambda service received the correct command/input. Probably ok though as we don't want to reconfigure this whole test if we can avoid it. 

Could this be fragile? 🤔 Maybe not because we're at a very high level and `invokeApi` isn't likely to make more calls to the lambda service in the future...?

File: packages/api/ecs/async-operation/package.json [hunk 4]
```diff
@@ -25,7 +25,7 @@
     "@cumulus/db": "18.2.0",
     "@cumulus/es-client": "18.2.0",
     "@cumulus/logger": "18.2.0",
-    "aws-sdk": "^2.1492.0",
+    "@aws-sdk/client-lambda": "^3.529.1",
```
Comment (npauzenga) [inline packages/api/ecs/async-operation/package.json:28]: ohh nice 😄. We should definitely start removing aws-sdk when we spot it and it's no longer used in the package.

File: packages/api/package.json [hunk 5]
```diff
@@ -8,7 +8,7 @@
   },
   "scripts": {
     "build-lambda-zips": "for x in $(ls dist); do (cd dist/${x} && rm -f lambda.zip && find . | xargs node ../../../../bin/zip.js lambda.zip && echo zipped ${x}); done",
-    "build": "../../node_modules/.bin/webpack",
+    "build": "node --max-old-space-size=4096 ../../node_modules/.bin/webpack",
```
Comment (npauzenga) [inline packages/api/package.json:11]: What does this do?

File: packages/api/tests/lib/rules/test-rulesHelpers.js [hunk 6]
```diff
@@ -835,15 +842,10 @@ test.serial('deleteRuleResources correctly deletes resources for kinesis rule',
...
-        promise: () => Promise.resolve(),
-      }),
-      removePermission: () => ({
-        promise: () => Promise.resolve(),
-      }),
-    });
+  const lambdaMock = mockClient(awsServices.lambda());
+  lambdaMock.onAnyCommand().rejects();
+  lambdaMock.on(RemovePermissionCommand).resolves({});
+  lambdaMock.on(AddPermissionCommand).resolves({});
```
Comment (npauzenga) [inline packages/api/tests/lib/rules/test-rulesHelpers.js:848]: nit: these used to resolve to undefined instead of an empty object.

File: packages/api/tests/lib/rules/test-rulesHelpers.js [hunk 7]
```diff
@@ -1318,6 +1319,9 @@ test.serial('deleteSnsTrigger throws more detailed ResourceNotFoundError', async
     state: 'ENABLED',
   }));
 
+  const lambdaStub = sinon.stub(awsServices.lambda(), 'send');
+  lambdaStub.throws(error);
```
Comment (npauzenga) [inline packages/api/tests/lib/rules/test-rulesHelpers.js:1323]: Different pattern but based on what we're testing I think it makes sense to stub the whole`send`.

File: packages/api/tests/lib/rules/test-rulesHelpers.js [hunk 8]
```diff
@@ -1712,8 +1705,7 @@ test.serial('Creating an enabled SNS rule creates an event source mapping', asyn
     state: 'ENABLED',
   });
   const subscribeSpy = sinon.spy(awsServices.sns(), 'subscribe');
-  const addPermissionSpy = sinon.spy(awsServices.lambda(), 'addPermission');
-
+  const lambdaMock = sinon.spy(awsServices.lambda(), 'send');
```
Comment (npauzenga) [inline packages/api/tests/lib/rules/test-rulesHelpers.js:1708]: Hmm we might want to be more specific here and listen for `addPermission`?

File: packages/api/webpack.config.js [hunk 9]
```diff
@@ -81,6 +81,7 @@ module.exports = {
   },
   externals: [
     'aws-sdk',
+    '@aws-sdk',
```
Comment (npauzenga) [inline packages/api/webpack.config.js:84]: Is the `@aws-sdk` the v3 version of `aws-sdk`? Adding to `externals` reduces our webpack size somehow?

File: packages/async-operations/package.json [hunk 10]
```diff
@@ -30,7 +30,7 @@
   "license": "Apache-2.0",
   "dependencies": {
     "@aws-sdk/client-ecs": "^3.477.0",
-    "@aws-sdk/client-lambda": "^3.447.0",
+    "@aws-sdk/client-lambda": "^3.529.1",
```
Comment (npauzenga) [inline packages/async-operations/package.json:33]: For the initial passes we were using the same 3.477 version but a decent amount of time has passed. It might be worth making it an AC on these tickets to go in and update the package you're working on to the most recent.

File: tf-modules/ingest/discover-granules-task.tf [hunk 11]
```diff
@@ -7,7 +7,7 @@ resource "aws_lambda_function" "discover_granules_task" {
   role             = var.lambda_processing_role_arn
   runtime          = "nodejs16.x"
   timeout          = lookup(var.lambda_timeouts, "DiscoverGranules", 300)
-  memory_size      = lookup(var.lambda_memory_sizes, "DiscoverGranules", 512)
+  memory_size      = lookup(var.lambda_memory_sizes, "DiscoverGranules", 1024)
```
Comment (npauzenga) [inline tf-modules/ingest/discover-granules-task.tf:10]: Same comment as above. Might be good to get this in sooner than later.

**Review by npauzenga — CHANGES_REQUESTED** [review]
Looks great @Jkovarik! A couple comments, mostly clarifying questions 😄

File: example/cumulus-tf/variables.tf (hunk 1 above)
Comment (npauzenga) [reply example/cumulus-tf/variables.tf:427]: Actually, this is a different Lambda than I was thinking. We need to increase for both Discover and Distribution?

Comment (Jkovarik) [reply example/cumulus-tf/variables.tf:427]: Yes, distribution is likely due to size bloat in the API package, although I can't prove it.   It didn't run *at all* without this increase.    We can talk more/demo if you'd like.

File: example/package.json (hunk 2 above)
Comment (Jkovarik) [reply example/package.json:75]: I'm not sure that it matters, functionally as this isn't a deployed package.    Happy to move them if we think that's right.

File: packages/api-client/tests/test-CumulusApiClient.js (hunk 3 above)
Comment (Jkovarik) [reply packages/api-client/tests/test-CumulusApiClient.js:29]: Right, that last thing, structurally the api-client package shouldn't *ever* be doing anything other than a direct invocation of the API lambda, this feels fairly safe.

File: packages/api/package.json (hunk 5 above)
Comment (Jkovarik) [reply packages/api/package.json:11]: This increases the memory available to webpack from the default to 4GB.    This was needed as (I believe) increased dependency size was causing the process to consume more memory than the default, resulting in failures on CI (but not local).

File: packages/api/tests/lib/rules/test-rulesHelpers.js (hunk 6 above)
Comment (Jkovarik) [reply packages/api/tests/lib/rules/test-rulesHelpers.js:848]: Fair - will address.

File: packages/api/tests/lib/rules/test-rulesHelpers.js (hunk 8 above)
Comment (Jkovarik) [reply packages/api/tests/lib/rules/test-rulesHelpers.js:1708]: Fair point, will address

File: packages/api/webpack.config.js (hunk 9 above)
Comment (Jkovarik) [reply packages/api/webpack.config.js:84]: It does, in that we're not including it as a top level dependency.   As noted yesterday, this needs removed until we upgrade to node v18

File: packages/api/tests/lib/rules/test-rulesHelpers.js (hunk 6 above)
Comment (Jkovarik) [reply packages/api/tests/lib/rules/test-rulesHelpers.js:848]: 1d3b0db02fc

File: packages/api/tests/lib/rules/test-rulesHelpers.js (hunk 8 above)
Comment (Jkovarik) [reply packages/api/tests/lib/rules/test-rulesHelpers.js]: 91a96be9c4b

**Comment by Jkovarik** [comment]
@npauzenga  Thanks for the review, I think I've addressed your comments.

File: example/cumulus-tf/variables.tf (hunk 1 above)
Comment (npauzenga) [reply example/cumulus-tf/variables.tf:427]: I don't know that we need to 100% prove that. I think it's fine to bump it up.

**Review by npauzenga — APPROVED** [review]
Thanks @Jkovarik I think this looks good assuming the memory allocation changes get around the errors we're seeing.
